README
CGI::ExtDirect
==============

    This module provides an RPC::ExtDirect gateway implementation for CGI
    compatible Web server environments. This includes both traditional CGI
    scripts that start up anew for each HTTP request, as well as more modern
    CGI environments in which a script is started once and then persists
    through the lifetime of a server process.

    CGI::ExtDirect can be used wth Perl versions 5.6 and newer with many Web
    servers; it was tested successfully with Apache/mod_perl, pure Perl
    server based on HTTP::Server::Simple (RPC::ExtDirect::Server), and
    various other HTTP server environments.

    If you are not familiar with Ext.Direct, more information can be found
    in RPC::ExtDirect::Intro.

INSTALLATION

    To install this module type the following:

        perl Makefile.PL
        make && make test
        make install

EXAMPLES

    See included Ext JS examples for ideas on what Ext.Direct is and how to
    use it in CGI applications. The examples are not installed along with
    the CGI::ExtDirect module, and are only available in the `examples/'
    directory of the CPAN distribution.

    To run examples type the following in the CGI::ExtDirect tarball
    directory:

        cd examples
        perl p5httpd

    Note that these examples do not require CGI::ExtDirect to be installed
    so you can try them beforehand. That said, CGI::ExtDirect depends on
    RPC::ExtDirect being available in `@INC' so if you don't want to install
    either module, unpack RPC::ExtDirect and CGI::ExtDirect tarballs to the
    same directory and use `$PERL5LIB' to point to RPC::ExtDirect location:

        cd examples
        PERL5LIB=../../RPC-ExtDirect-3.xx/lib perl p5httpd

COPYRIGHT AND LICENSE

    Copyright (c) 2011-2015 Alex Tokarev <tokarev@cpan.org>.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic.

    Included Ext JS examples are copyright (c) 2011, Sencha Inc. Example
    code is used and distributed under GPL 3.0 license as provided by Sencha
    Inc. See http://www.sencha.com/license. Ext JS is available for download
    at http://www.sencha.com/products/extjs/



Changes
3.24  Wed Oct 12 00:54:25 PDT 2016
    - RPC::ExtDirect dependency bumped to 3.24
    - CGI::ExtDirect version bumped to 3.24 to match the core module
    - No code changes since last dev version

3.21_001 Mon Jun 15 22:04:46 PDT 2015
    - Require CGI 4.20+ if it's over 4.0; trying to work around Win32 test failures

3.20  Tue Mar 31 21:46:14 PDT 2015
    - Updated to support call metadata feature in RPC::ExtDirect 3.20
    - Added unit tests
    - RPC::ExtDirect depdendency bumped to 3.20
    - CGI::Test dependency bumped to 1.100

3.12  Sun Jan 18 12:10:53 PST 2015
    - Updated CDN links in examples to Ext JS 4.2.1 instead of
      outdated and no longer supported 4.0.2a.
    - No code changes.

3.11  Mon Oct 27 17:41:48 PDT 2014
    - Suppress list context warning in CGI.pm 4.08+
    - RPC::ExtDirect dependency bumped to 3.02
    - CGI::Test dependency bumped to 1.000

3.10  Fri Sep  5 22:46:40 PDT 2014
    - Windows compatibility
    - CGI::Test dependency bumped to 0.50

3.01  Fri Jun 20 19:46:26 PDT 2014
    - Fixed test failures due to missing heredoc closing tags
      in test scripts
    - Some documentation touch-ups

3.00  Thu Jun 12 18:19:37 PDT 2014
    - Updated to accommodate for the changes in RPC::ExtDirect 3.0
    - Many bug fixes, the bulk of the test data moved to the core
    - More documentation

2.03  Tue Jun 10 20:02:18 PDT 2014
    - Fixed a bug in CGI header handling that made HTTP response
      output to be mangled sometimes with Perls 5.18+

2.02  Thu Jun 21 10:27:33 2012
    - Fixed a bug in t/05_env.t

2.01  Wed Jun 20 18:37:54 2012
    - Minor documentation tweaks.

2.00  Mon Jun 18 14:15:11 2012
    - Updated code, documentation and test suite to accommodate
      for new features in RPC::ExtDirect 2.00.

1.12  Thu Jun  7 13:41:33 2012
    - Fixed a bug in handling uploaded files: CGI.pm below 3.41
      did not have "handle" method in its "lightweight file handle"
      objects, which CGI::ExtDirect relied on.
      Somehow this bug only manifested itself in Perl 5.10.0.

    - Fixed Makefile.PL to produce BUILD_REQUIRES only for 
      ExtUtils::MakeMaker that support it (>=6.55).

    - Bumped RPC::Dependency to freshly released 1.30.

1.11  Mon May 20 18:54:05 2012
    - Bumped CGI::Test dependency to version 0.3 which should not
      fail its tests anymore. Moved it to BUILD_REQUIRES section.

      Bumped RPC::ExtDirect dependency version to 1.21.

      Added repository and bugtracker attributes to Makefile.PL.

      Updated README a bit.

      No code changes this time.

1.10  Mon Oct 17 00:21:42 2011
    - Fixed a (small) bug: extType parameter in form-based remoting
      calls was not caught and cleaned out before passing data hash
      to Method for execution.

      Removed CGI::Test notice from README; I am now maintainer of
      that module and CPAN distribution should be straightened
      already.

1.01  Tue Oct  4 21:35:06 2011
    - Bumped required RPC::ExtDirect version to 1.20.

1.00  Sat Oct  1 20:59:11 2011
	- original version.


t/lib/RPC/ExtDirect/Test/Env.pm
package RPC::ExtDirect::Test::Env;

use strict;
use warnings;
no  warnings 'uninitialized';

use RPC::ExtDirect class => 'Env';

sub http_list : ExtDirect(0) {
    my ($class, $env) = @_;

    my @list = sort $env->http();

    return [ @list ];
}

sub http_header : ExtDirect(1) {
    my ($class, $header, $env) = @_;

    return $env->http($header);
}

sub param_list : ExtDirect(0) {
    my ($class, $env) = @_;

    my @list = sort $env->param();

    return [ @list ];
}

sub param_get : ExtDirect(1) {
    my ($class, $name, $env) = @_;

    return $env->param($name);
}

sub cookie_list : ExtDirect(0) {
    my ($class, $env) = @_;

    my @cookies = sort $env->cookie();

    return [ @cookies ];
}

sub cookie_get : ExtDirect(1) {
    my ($class, $name, $env) = @_;

    return $env->cookie($name);
}

1;



t/lib/RPC/ExtDirect/Test/Util/CGI.pm
package RPC::ExtDirect::Test::Util::CGI;

use strict;
use warnings;
no  warnings 'uninitialized';

use Test::More;

use CGI::Test ();       # No need to import ok() from CGI::Test
use CGI::Test::Input::URL ();
use CGI::Test::Input::Multipart ();

use RPC::ExtDirect::Test::Util;

use base 'Exporter';

our @EXPORT = qw/
    run_tests
/;

our @EXPORT_OK = qw/
    raw_post
    form_post
    form_upload
/;

use constant WINDOWS => eval { $^O =~ /Win32|cygwin/ };

### EXPORTED PUBLIC PACKAGE SUBROUTINE ###
#
# Run the test battery from the passed definitions
#

sub run_tests {
    my ($tests, @run_only) = @_;
    
    my $cmp_pkg   = 'RPC::ExtDirect::Test::Util';
    my $num_tests = @run_only || @$tests;
    
    plan tests => 4 * $num_tests;

    TEST:
    for my $test ( @$tests ) {
        my $name   = $test->{name};
        my $config = $test->{config} || {};
        my $input  = $test->{input};
        my $output = $test->{output};
        
        next TEST if @run_only && !grep { lc $name eq lc $_ } @run_only;
        
        my $ct = CGI::Test->new(
            -base_url => 'http://localhost/cgi-bin',
            -cgi_dir  => 't/cgi-bin',
            %$config,
        );
    
        # CGI tests have the config hardcoded in the scripts
        my $url           = $ct->base_uri
                          . $input->{cgi_url}
                          . ( WINDOWS ? '.bat' : '' );
        my $method        = $input->{method};
        my $input_content = $input->{cgi_content} || $input->{content};
        
        my $req = prepare_input 'CGI', $input_content;
        my $page = $ct->$method($url, $req);

        if ( ok $page, "$name not empty" ) {
            my $want_type = $output->{content_type};
            my $have_type = $page->content_type();
            
            like $have_type, $want_type, "$name content type";
            
            my $want_status = $output->{status};
            my $have_status = $page->is_ok() ? 200 : $page->error_code();

            is $have_status, $want_status, "$name HTTP status";

            my $cmp_fn = $output->{comparator};
            my $want   = $output->{cgi_content} || $output->{content};
            my $have   = $page->raw_content();
            
            $cmp_pkg->$cmp_fn($have, $want, "$name content")
                or diag explain "Page: ", $page;

            $page->delete();
        };
    };
}

### NON EXPORTED PUBLIC PACKAGE SUBROUTINE ###
#
# Return a new CGI::Test::Input object for a raw POST call
#

sub raw_post {
    # This can be called either as a class method, or a plain sub
    shift if $_[0] eq __PACKAGE__;
    
    my ($url, $input) = @_;

    my $cgi_input = CGI::Test::Input::URL->new();
    $cgi_input->set_raw_data($input);
    $cgi_input->set_mime_type('application/json');

    return $cgi_input;
}

### NON EXPORTED PUBLIC PACKAGE SUBROUTINE ###
#
# Return a new CGI::Test::Input oject for a form call
#

sub form_post {
    # This can be called either as a class method, or a plain sub
    shift if $_[0] eq __PACKAGE__;

    my ($url, %fields) = @_;

    my $cgi_input = CGI::Test::Input::URL->new();
    for my $field ( keys %fields ) {
        my $value = $fields{ $field };
        $cgi_input->add_field($field, $value);
    };

    return $cgi_input;
}

### NON EXPORTED PUBLIC PACKAGE SUBROUTINE ###
#
# Return a new CGI::Test::Input object for a form call
# with file uploads
#

sub form_upload {
    # This can be called either as a class method, or a plain sub
    shift if $_[0] eq __PACKAGE__;

    my ($url, $files, %fields) = @_;

    my $cgi_input = CGI::Test::Input::Multipart->new();

    for my $field ( keys %fields ) {
        my $value = $fields{ $field };
        $cgi_input->add_field($field, $value);
    };

    for my $file ( @$files ) {
        $cgi_input->add_file_now("upload", "t/data/cgi-data/$file");
    };

    return $cgi_input;
}


1;


lib/CGI/ExtDirect.pm
package CGI::ExtDirect;

use strict;
use warnings;
no  warnings 'uninitialized';       ## no critic

use Carp;
use IO::Handle;
use File::Basename qw(basename);

use RPC::ExtDirect::Util ();
use RPC::ExtDirect::Config;
use RPC::ExtDirect::API;
use RPC::ExtDirect;

#
# This module is not compatible with RPC::ExtDirect < 3.0
#

die __PACKAGE__." requires RPC::ExtDirect 3.0+"
    if $RPC::ExtDirect::VERSION lt '3.0';

### PACKAGE GLOBAL VARIABLE ###
#
# Version of this module.
#

our $VERSION = '3.21_001';

### PUBLIC CLASS METHOD (CONSTRUCTOR) ###
#
# Instantiate a new CGI::ExtDirect object
#

sub new {
    my $class = shift;

    my %arg = @_ == 1 && 'HASH' eq ref $_[0] ? %{ $_[0] }
            :                                  @_
            ;
    
    my $api    = delete $arg{api}    || RPC::ExtDirect->get_api();
    my $config = delete $arg{config} || $api->config;
    
    # We need a CGI object for input processing
    my $cgi = $arg{cgi} || do { require CGI; new CGI };

    # Debug flag defaults to off
    $config->debug( $arg{debug} ) if exists $arg{debug};

    my $self = bless {
        config  => $config,
        api_obj => $api,
        cgi     => $cgi,
        %arg,
    }, $class;

    return $self;
}

### PUBLIC INSTANCE METHOD ###
#
# Returns API definition for ExtDirect, along with headers
#

sub api {
    my ($self, @headers) = @_;

    # Get the API JavaScript
    my $js = eval {
        $self->api_obj->get_remoting_api(
            config => $self->config,
            env    => $self->cgi,
        )
    };

    # If JS API call failed, return error headers
    # What exactly went wrong is not too relevant here
    return $self->error_headers(@headers) if $@;

    # If API call succeed, return application/javascript with 200 OK
    my $content_type = 'application/javascript';
    my $http_status  = '200 OK';

    # And we need content length, too (in octets)
    my $content_length = do { use bytes; length $js; };

    # Munge the headers passed on us
    my @real_headers = $self->_munge_headers($content_type,
                                             $http_status,
                                             $content_length,
                                             @headers);

    # Finally, compile HTTP response
    my $response = $self->cgi->header(@real_headers) .
                   $js;

    return $response;
}

### PUBLIC INSTANCE METHOD ###
#
# Routes the action request and returns HTTP response with headers
#

sub route {
    my ($self, @headers) = @_;

    # If any but POST method is used, just throw an error
    return $self->error_headers(@headers)
        if $self->cgi->request_method() ne 'POST';

    # Try to distinguish between raw POST and form call (Ugh)
    my $router_input = $self->_extract_post_data();

    # When extraction fails, undef is returned
    return $self->error_headers(@headers)
        unless defined $router_input;
    
    my $config       = $self->config;
    my $api          = $self->api_obj;
    my $router_class = $config->router_class;
    
    eval "require $router_class";
    
    my $router = $router_class->new(
        config => $config,
        api    => $api,
    );

    # Routing requests is safe (Router won't croak under torture)
    my $result = $router->route($router_input, $self->cgi);

    my ($content_type, $http_body, $content_length);

    $content_type   = $result->[1]->[1];
    $content_length = $result->[1]->[3];
    $http_body      = $result->[2]->[0];
    my $http_status = '200 OK';

    # Munge the headers passed on us
    my @real_headers = $self->_munge_headers($content_type,
                                             $http_status,
                                             $content_length,
                                             @headers);

    # Finally, compile HTTP response
    my $response = $self->cgi->header(@real_headers) .
                   $http_body;

    return $response;
}

### PUBLIC INSTANCE METHOD ###
#
# Queries Event providers for events, returning serialized stream.
#

sub poll {
    my ($self, @headers) = @_;

    # Only GET and POST methods are supported for polling
    return $self->error_headers(@headers)
        if $self->cgi->request_method() !~ / \A (GET|POST) \z /xms;
    
    my $config         = $self->config;
    my $api            = $self->api_obj;
    my $provider_class = $config->eventprovider_class;
    
    eval "require $provider_class";
    
    my $provider = $provider_class->new(
        config => $config,
        api    => $api,
    );

    # Polling for Events is safe
    my $http_body = $provider->poll($self->cgi);

    # Gather variables for HTTP response
    my $content_type = 'application/json';
    my $http_status  = '200 OK';

    # And we need content length, too (in octets)
    my $content_length = do { use bytes; length $http_body; };

    # Munge the headers passed on us
    my @real_headers = $self->_munge_headers($content_type,
                                             $http_status,
                                             $content_length,
                                             @headers);

    # Finally, compile HTTP response
    my $response = $self->cgi->header(@real_headers) .
                   $http_body;

    return $response;
}

### PUBLIC INSTANCE METHOD ###
#
# Returns error HTTP header string. There is not much sense in
# returning HTTP body as well since Ext.Direct calls are automated
# and there is nobody to see error messages anyway.
#

sub error_headers {
    my ($self, @headers) = @_;

    # Get ourselves a set of brand new CGI headers
    my @cgi_headers = $self->_munge_headers('text/html',
                                            '500 Internal Server Error',
                                            0,
                                            @headers);

    return $self->cgi->header(@cgi_headers);
}

### PUBLIC INSTANCE METHODS ###
#
# Read-write accessors
#

RPC::ExtDirect::Util::Accessor->mk_accessors(
    simple => [qw/ config api_obj cgi /],
);

############## PRIVATE METHODS BELOW ##############

### PRIVATE INSTANCE METHOD ###
#
# Munges CGI headers so that they become what we need
#

sub _munge_headers {
    my ($self, $content_type, $http_status,
               $content_length, @headers) = @_;

    # Default charset is UTF-8
    my $charset = 'utf-8';

    # First form is no additional headers passed on us, the easy one
    # Second form includes only one parameter and that's content type
    # Third form includes both content type and HTTP status
    # The last form is a hash of headers but we'd better check anyway
    return (
             '-type'           => $content_type,
             '-status'         => $http_status,
             '-charset'        => $charset,
             '-content_length' => $content_length,
           )
        if  @headers == 0 || @headers == 1 ||
           (@headers == 2 && $headers[0] !~ / \A - /msx) ||
           (@headers > 2 && ((@headers % 2) != 0));

    # Finally we've got a hash of header parameters
    my %cgi_headers = @headers;

    # Interesting are the headers we need to deal with
    my %interesting_item = (
        '-type'           => qr/ \A -? (content [-_])? type \z /ixms,
        '-status'         => qr/ \A -? status \z               /ixms,
        '-charset'        => qr/ \A -? charset \z              /ixms,
        '-content_length' => qr/ \A -? content [-_] length \z  /ixms,
        '-nph'            => qr/ \A -? nph \z                  /ixms,
    );

    # Normalize them headers we need, don't touch the others
    HEADER_ITEM:
    for my $item ( keys %interesting_item ) {
        my $pattern = $interesting_item{ $item };

        # First find all occurences of the interesting item
        my @found_items = grep { /$pattern/ } keys %cgi_headers;
        next HEADER_ITEM unless @found_items;

        # Then take the *first* value -- we don't care about duplicates
        # and they should not have happened anyway, so there
        my $value = $cgi_headers{ $found_items[0] };

        # Delete all occurences of the item in question
        delete @cgi_headers{ @found_items };

        # Finally, place normalized item back in hash
        $cgi_headers{ $item } = $value;
    };

    # Make sure we have the required headers
    $cgi_headers{'-type'} = $content_type
        unless exists $cgi_headers{ '-type' };
    
    $cgi_headers{'-status'} = $http_status
        unless exists $cgi_headers{ '-status' };
    
    # Content-length we force
    $cgi_headers{'-content_length'} = $content_length;

    # If they passed charset, then they probably know what they're doing
    $cgi_headers{ '-charset' } = $charset
        unless exists $cgi_headers{ '-charset' };

    # Defang CGI.pm's interface idiosyncracies by ensuring that
    # a header starting with a dash always comes first. Otherwise
    # the hash key randomizer introduced in Perl 5.18 may screw up
    # for us by placing a header with no dash in the first place,
    # making CGI->header() think that it has been fed the first argument
    # form header('content/type', 'HTTP status') instead of the hash
    # form. This leads to CGI::ExtDirect returning a HTTP status line
    # like "HTTP/1.1 1" instead of "HTTP/1.1 200 OK" *sometimes*.
    # Dang.
    return (
        '-type' => delete $cgi_headers{ '-type' },
         %cgi_headers,
    );
}

### PRIVATE INSTANCE METHOD ###
#
# Deals with intricacies of POST-fu and returns something suitable to
# feed to the Router (string or hashref, really). Or undef if something
# goes too wrong to recover.

my @STANDARD_KEYWORDS
    = qw(action method extAction extMethod extTID extUpload extType); 
my %STANDARD_KEYWORD = map { $_ => 1 } @STANDARD_KEYWORDS;

sub _extract_post_data {
    my ($self) = @_;

    # We need CGI object here real bad
    my $cgi = $self->cgi;

    # The smartest way to tell if a form was submitted that *I* know of
    # is to look for 'extAction' and 'extMethod' keywords in CGI params.
    my %keyword = map { $_ => 1 } $cgi->param();
    my $is_form = exists $keyword{ extAction } &&
                  exists $keyword{ extMethod };

    # If form is not involved, it's easy: just return POSTDATA (or undef)
    if ( !$is_form ) {
        my $postdata = $cgi->param('POSTDATA');
        return $postdata ne '' ? $postdata
               :                 undef
               ;
    };

    # If any files are attached, extUpload will contain 'true'
    my $has_uploads = $cgi->param('extUpload') eq 'true';

    # Here file uploads data is stored
    my @_uploads = ();

    # This is to suppress a really annoying warning in CGI.pm 4.08+.
    # I am perfectly aware of what the list context is and how to
    # use it, thank you very much. :/
    local $CGI::LIST_CONTEXT_WARN = 0;

    # Now if the form IS involved, it gets a little bit complicated
    PARAM:
    for my $param ( keys %keyword ) {
        # Defang CGI's idiosyncratic way of returning multi-valued params
        my @values = $cgi->param( $param );
        $keyword{ $param } = @values == 0 ? undef
                           : @values == 1 ? $values[0]
                           :                [ @values ]
                           ;

        # Try to see if $param is a field with associated file upload
        # Skip the standard ones first, of course
        next PARAM if $STANDARD_KEYWORD{ $param } || !$has_uploads;

        # Look for file uploads in this field
        my @field_uploads = $self->_parse_uploads($cgi, $param);

        # Found some, add them to the general stash and kill the field
        if ( @field_uploads ) {
            push @_uploads, @field_uploads;
            delete $keyword{ $param };
        };
    };

    # Metadata is JSON encoded; decode_metadata lives by side effects!
    if ( exists $keyword{metadata} ) {
        RPC::ExtDirect::Util::decode_metadata($self, \%keyword);
    }

    # Remove extType because it's meaningless later on
    delete $keyword{ extType };

    # Fix up the TID so that it comes as a number (JavaScript is picky)
    $keyword{ extTID } += 0 if exists $keyword{ extTID };

    # Now add files to hash, if any
    $keyword{ '_uploads' } = \@_uploads if @_uploads;

    return \%keyword;
}

### PRIVATE INSTANCE METHOD ###
#
# Parses CGI form input field looking for file uploads
#

sub _parse_uploads {
    my ($self, $cgi, $param) = @_;

    # CGI returns "lightweight file handles", or undef
    my @file_handles = $cgi->upload($param);

    # Empty list means no uploads for this field
    return unless grep { defined $_ } @file_handles;

    # Despite what CGI documentation says, the values returned
    # as "file names" are actually some kind of key handles
    my @file_keys = $cgi->param($param);

    # Here file uploads get collected
    my @uploads = ();

    # Collect the info we need to repackage it in a consistent way
    FILE:
    for my $key ( @file_keys ) {
        # First take a closer look at this "blah-blah handle"
        my $file_handle = shift @file_handles;

        # undef would mean there was an upload error (timeout perhaps)
        # Following HTTP POST logic, when one upload breaks, that
        # would mean all subsequent uploads in this POST are also
        # broken.
        # We can't recover from that so just stop trying.
        last FILE unless defined $file_handle;

        # In CGI.pm < 3.41, "lightweight handle" object doesn't support
        # returning IO::Handle so we do it manually to avoid problems
        my $io_handle = IO::Handle->new_from_fd(fileno $file_handle, '<');

        # We also need a lot of info about the file (if provided)
        my $upload_info = $cgi->uploadInfo($key);
        my $temp_file   = $cgi->tmpFileName($key);
        my $file_type   = $upload_info->{'Content-Type'};
        my $file_name   = $self->_get_file_name($upload_info);
        my $file_size   = $self->_get_file_size($io_handle);
        my $base_name   = basename($file_name);

        # Now instead of a "blah-blah handle" we have a normalized hashref
        push @uploads, {
            type     => $file_type,
            size     => $file_size,
            path     => $temp_file,
            handle   => $io_handle,
            basename => $base_name,
            filename => $file_name,
        };
    };

    return @uploads;
}

### PRIVATE INSTANCE METHOD ###
#
# Tries hard to extract file name from multipart form guts
#

sub _get_file_name {
    my ($self, $upload_info) = @_;

    # Pluck file name from Content-Disposition string
    my ($file_name)
        = $upload_info->{'Content-Disposition'} =~ /filename="(.*?)"/;

    # URL unescape it
    $file_name =~ s/%([\dA-Fa-f]{2})/pack("C", hex $1)/eg;

    return $file_name;
}

### PRIVATE INSTANCE METHOD ###
#
# Enquiries IO::Handle supplied by CGI for file size
#

sub _get_file_size {
    my ($self, $handle) = @_;

    # Fall through in case $handle is invalid
    return unless $handle;

    return ($handle->stat)[7];
}

1;


Makefile.PL
use 5.006000;
use ExtUtils::MakeMaker;

# Add the `devtest` target to run regression and POD tests in one go
sub MY::postamble {
    return <<'END';
devtest :
	REGRESSION_TESTS=1 POD_TESTS=1 $(MAKE) test

END
}

# Override `disttest` so it would behave as `devtest`
sub MY::dist_test {
    return <<'END';
disttest : distdir
	cd $(DISTVNAME) && $(ABSPERLRUN) Makefile.PL
	cd $(DISTVNAME) && $(MAKE) $(PASTHRU)
	cd $(DISTVNAME) && $(MAKE) devtest $(PASTHRU)

END
}

my $MM_VERSION  = $ExtUtils::MakeMaker::VERSION;
my $github_repo = 'https://github.com/nohuhu/CGI-ExtDirect';

my %CORE_REQ = (
    # 5.20 warns that CGI is going to be removed from
    # perl core in 5.22, so we require it here
    'CGI'            => 0,
    'RPC::ExtDirect' => '3.20',
);

# CGI.pm was in a flux between 4.00 and 4.20; if we have 4.00+ let's have latter one
eval {
    require CGI;
    $CORE_REQ{CGI} = '4.20' if $CGI::VERSION >= '4.00';
};

my %TEST_REQ = (
    'Test::More' => '0.82', # for explain()
    'CGI::Test'  => '1.100',
);

WriteMakefile(
    NAME         => 'CGI::ExtDirect',
    VERSION_FROM => 'lib/CGI/ExtDirect.pm',
    
    ABSTRACT     => 'RPC::ExtDirect gateway for CGI', 
    AUTHOR       => 'Alex Tokarev <tokarev@cpan.org>',
    LICENSE      => 'perl',
    
    ($MM_VERSION >= 6.64
        ? (
            TEST_REQUIRES => \%TEST_REQ,
            PREREQ_PM     => \%CORE_REQ,
        )
        : (
            ($MM_VERSION >= 6.5503
                ? ( 
                    BUILD_REQUIRES => \%TEST_REQ,
                    PREREQ_PM      => \%CORE_REQ,
                )
                : (
                    PREREQ_PM => {
                        %TEST_REQ,
                        %CORE_REQ,
                    },
                )
            ),
        ),
    ),

    ($MM_VERSION >= 6.48 ? ( MIN_PERL_VERSION => 5.006000 ) : ()),

    ($MM_VERSION >= 6.46
        ? ( META_MERGE => {
                resources   => {
                    bugtracker  => "$github_repo/issues",
                    repository  =>  $github_repo,
                },
            },
        )
        : ()
    ),
);

